tests/test_bootstrap.py

```diff
@@ -0,0 +1,545 @@
+    @mock.patch("pythonforandroid.bootstrap.sh.rm")
+    @mock.patch("pythonforandroid.bootstrap.sh.cp")
+    def test_run_distribute(
+        self,
+        mock_sh_cp,
+        mock_sh_rm,
+        mock_listdir,
+        mock_chdir,
+        mock_ensure_dir,
+        mock_strip_libraries,
+        mock_create_python_bundle,
+        mock_open_dist_files,
+        mock_open_sdl2_files,
+        mock_open_webview_files,
+        mock_open_service_only_files,
+    ):
+        """
+        A test for any overwritten method of
+        `~pythonforandroid.bootstrap.Bootstrap.run_distribute`. Here we mock
+        any file/dir operation that it could slow down our tests, and there is
+        a lot to mock, because the `run_distribute` method it should take care
+        of prepare all compiled files to generate the final `apk`. The targets
+        of this test will be:
+
+            - :meth:`~pythonforandroid.bootstraps.sdl2.BootstrapSdl2
+              .run_distribute`
+            - :meth:`~pythonforandroid.bootstraps.service_only
+              .ServiceOnlyBootstrap.run_distribute`
+            - :meth:`~pythonforandroid.bootstraps.webview.WebViewBootstrap
+               .run_distribute`
+            - :meth:`~pythonforandroid.bootstraps.empty.EmptyBootstrap.
+              run_distribute`
+
+        Here we will tests all those methods that are specific for each class.
+        """
+        # prepare bootstrap and distribution
+        bs = Bootstrap().get_bootstrap(self.bootstrap_name, self.ctx)
+        bs.build_dir = bs.get_build_dir()
+        self.setUp_distribution_with_bootstrap(bs)
+
+        self.ctx.hostpython = "/some/fake/hostpython3"
+        self.ctx.python_recipe = Recipe.get_recipe("python3", self.ctx)
+        self.ctx.python_modules = ["requests"]
+        self.ctx.archs = [ArchARMv7_a(self.ctx)]
+
+        bs.run_distribute()
+
+        mock_open_dist_files.assert_called_once_with("dist_info.json", "w")
+        mock_open_bootstraps = {
+            "sdl2": mock_open_sdl2_files,
+            "webview": mock_open_webview_files,
+            "service_only": mock_open_service_only_files,
+        }
+        expected_open_calls = {
+            "sdl2": [
+                mock.call("local.properties", "w"),
+                mock.call("blacklist.txt", "a"),
+            ],
+            "webview": [mock.call("local.properties", "w")],
+            "service_only": [mock.call("local.properties", "w")],
+        }
+        mock_open_bs = mock_open_bootstraps[self.bootstrap_name]
+        # test that the expected calls has been called
+        for expected_call in expected_open_calls[self.bootstrap_name]:
+            self.assertIn(expected_call, mock_open_bs.call_args_list)
+        # test that the write function has been called with the expected args
+        self.assertIn(
+            mock.call().__enter__().write("sdk.dir=/opt/android/android-sdk"),
+            mock_open_bs.mock_calls,
+        )
```

Comment:
@AndreMiras,
now we test that we actually we write the expected values...this (and the lines below) replaces the `read_data` thing that it was wrong :wink:

And **thanks for reviewing this**, your reviews allowed us to improve this pr a lot :heart:, now it looks much cleaner and we do more testing than the first posted version.

Reply:
That definitely makes more sense to me thanks! :clap: